src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +542,1177 @@ private void checkDataLength(int ... lengths) {
-                // Code below will write the remainder from 'in' to ibuffer
-            } else if (remainder > 0) {
-                // If a block or more was encrypted from 'buffer' only, but the
-                // rest of 'buffer' with 'in' could not construct a block, then
-                // put the rest of 'buffer' back into ibuffer.
-                ibuffer.reset();
-                ibuffer.write(buffer, blen, remainder);
-                // Code below will write the remainder from 'in' to ibuffer
+        /**
+         * Check for overlap. If the src and dst buffers are using shared data
+         * and if dst will overwrite src data before src can be processed.
+         * If so, make a copy to put the dst data in.
+         */
+        ByteBuffer overlapDetection(ByteBuffer src, ByteBuffer dst) {
+            if (src.isDirect() && dst.isDirect()) {
+                DirectBuffer dsrc = (DirectBuffer) src;
+                DirectBuffer ddst = (DirectBuffer) dst;
+
+                // Get the current memory address for the given ByteBuffers
+                long srcaddr = dsrc.address();
+                long dstaddr = ddst.address();
+
+                // Find the lowest attachment that is the base memory address
+                // of the shared memory for the src object
+                while (dsrc.attachment() != null) {
+                    srcaddr = ((DirectBuffer) dsrc.attachment()).address();
+                    dsrc = (DirectBuffer) dsrc.attachment();
+                }
+
+                // Find the lowest attachment that is the base memory address
+                // of the shared memory for the dst object
+                while (ddst.attachment() != null) {
+                    dstaddr = ((DirectBuffer) ddst.attachment()).address();
+                    ddst = (DirectBuffer) ddst.attachment();
+                }
+
+                // If the base addresses are not the same, there is no overlap
+                if (srcaddr != dstaddr) {
+                    return dst;
+                }
+                // At this point we know these objects share the same memory.
+                // This checks the starting position of the src and dst address
+                // for overlap.
+                // It uses the base address minus the passed object's address to
+                // get the offset from the base address, then add the position()
+                // from the passed object.  That gives up the true offset from
+                // the base address.  As long as the src side is >= the dst
+                // side, we are not in overlap.
+                if (((DirectBuffer) src).address() - srcaddr + src.position() >=
+                    ((DirectBuffer) dst).address() - dstaddr + dst.position()) {
+                    return dst;
+                }
+
+            } else if (!src.isDirect() && !dst.isDirect()) {
+                // if src is read only, then we need a copy
+                if (!src.isReadOnly()) {
```

Comment:
I assume this is again the case of !isDirect()&&!isReadOnly => hasArray()?

Reply:
hasArray() would not be appropriate here.  The code needs to get into this if, being a heap bytebuffer.  If I use hasArray() it will skip to the ending else and cause a test failure.  Having the isReadOnly checked inside the if(isDirect) causes it to create the necessary copy